File: app/src/main/java/top/easelink/lcg/ui/main/view/MainActivity.java
package top.easelink.lcg.ui.main.view;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.databinding.DataBindingUtil;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.lifecycle.ViewModelProviders;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.tabs.TabLayout;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.jetbrains.annotations.NotNull;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.EmptyStackException;
import java.util.List;
import java.util.Properties;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.support.HasSupportFragmentInjector;
import pub.devrel.easypermissions.EasyPermissions;
import top.easelink.framework.base.BaseActivity;
import top.easelink.framework.base.BaseFragment;
import top.easelink.lcg.BR;
import top.easelink.lcg.R;
import top.easelink.lcg.databinding.ActivityMainBinding;
import top.easelink.lcg.databinding.NavHeaderMainBinding;
import top.easelink.lcg.mta.EventHelperKt;
import top.easelink.lcg.ui.main.about.view.AboutFragment;
import top.easelink.lcg.ui.main.article.view.ArticleFragment;
import top.easelink.lcg.ui.main.articles.view.ArticlesFragment;
import top.easelink.lcg.ui.main.articles.view.FavoriteArticlesFragment;
import top.easelink.lcg.ui.main.articles.view.ForumArticlesFragment;
import top.easelink.lcg.ui.main.forumnav.view.ForumNavigationFragment;
import top.easelink.lcg.ui.main.me.view.MeFragment;
import top.easelink.lcg.ui.main.model.NewMessageEvent;
import top.easelink.lcg.ui.main.model.NotificationInfo;
import top.easelink.lcg.ui.main.model.OpenArticleEvent;
import top.easelink.lcg.ui.main.model.OpenForumEvent;
import top.easelink.lcg.ui.main.model.OpenNotificationsPageEvent;
import top.easelink.lcg.ui.main.model.TabModel;
import top.easelink.lcg.ui.main.viewmodel.MainViewModel;
import top.easelink.lcg.ui.search.view.SearchActivity;
import top.easelink.lcg.utils.ActivityUtils;
import top.easelink.lcg.utils.ToastUtilsKt;

import static top.easelink.lcg.mta.MTAConstantKt.EVENT_OPEN_FORUM;
import static top.easelink.lcg.mta.MTAConstantKt.PROP_FORUM_NAME;
import static top.easelink.lcg.utils.ActivityUtils.TAG_PREFIX;
import static top.easelink.lcg.utils.WebsiteConstant.APP_RELEASE_PAGE;
import static top.easelink.lcg.utils.WebsiteConstant.SEARCH_URL;
import static top.easelink.lcg.utils.WebsiteConstant.URL_KEY;

public class MainActivity
        extends BaseActivity<ActivityMainBinding, MainViewModel>
        implements HasSupportFragmentInjector,
        BottomNavigationView.OnNavigationItemSelectedListener,
        EasyPermissions.PermissionCallbacks {

    @Inject
    DispatchingAndroidInjector<Fragment> fragmentDispatchingAndroidInjector;
    private static WeakReference<MainActivity> mainActivityWeakReference;
    private DrawerLayout mDrawer;
    private Long lastBackPressed = 0L;

    @Override
    public int getBindingVariable() {
        return BR.viewModel;
    }

    @Override
    public int getLayoutId() {
        return R.layout.activity_main;
    }

    @Override
    public MainViewModel getViewModel() {
        return ViewModelProviders.of(this).get(MainViewModel.class);
    }

    @Override
    public void onBackPressed() {
        if (!mFragmentTags.empty()) {
            String tag = mFragmentTags.peek();
            if (onFragmentDetached(tag)) {
                mFragmentTags.pop();
                syncBottomViewNavItemState();
                return;
            } else {
                // check viewpager
                ViewPager viewPager = getViewDataBinding().mainViewPager;
                if (viewPager.isShown() && viewPager.getCurrentItem() != 0) {
                    viewPager.setCurrentItem(0, true);
                    return;
                }
            }
        }
        if (System.currentTimeMillis() - lastBackPressed > 2000) {
            Toast.makeText(this, R.string.app_exit_tip, Toast.LENGTH_SHORT).show();
            lastBackPressed = System.currentTimeMillis();
        } else {
            finish();
            System.exit(0);
        }
    }

    @Override
    public AndroidInjector<Fragment> supportFragmentInjector() {
        return fragmentDispatchingAndroidInjector;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EventBus.getDefault().register(this);
        mainActivityWeakReference = new WeakReference<>(this);
        setUp();
        requestPermissions();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (mDrawer != null) {
            mDrawer.setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

    private void lockDrawer() {
        if (mDrawer != null) {
            mDrawer.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED);
        }
    }

    /**
     * manage the bottom navigation view item selected state
     * depends on current top fragment
     */
    private void syncBottomViewNavItemState() {
        BottomNavigationView view = getViewDataBinding().bottomNavigation;
        try {
            String topFragment = mFragmentTags.peek();
            view.setOnNavigationItemSelectedListener(null);
            if (topFragment.startsWith("android")) {
                // means that current top fragment is the tab-layout
                view.setSelectedItemId(R.id.action_home);
                unlockDrawer();
            } else if (topFragment.endsWith(FavoriteArticlesFragment.class.getSimpleName())) {
                view.setSelectedItemId(R.id.action_favorite);
            } else if (topFragment.endsWith(ForumNavigationFragment.class.getSimpleName())) {
                view.setSelectedItemId(R.id.action_forum_navigation);
            }
        } catch (EmptyStackException ese) {
            view.setSelectedItemId(R.id.action_home);
        } finally {
            view.setOnNavigationItemSelectedListener(this);
        }
    }

    private void setUp() {
        mDrawer = getViewDataBinding().drawerView;
        Toolbar mToolbar = getViewDataBinding().toolbar;
        TabLayout mTabLayout = getViewDataBinding().mainTab;
        setSupportActionBar(mToolbar);
        ActionBarDrawerToggle mDrawerToggle = new ActionBarDrawerToggle(
                this,
                mDrawer,
                mToolbar,
                R.string.open_drawer,
                R.string.close_drawer) {
            @Override
            public void onDrawerClosed(View drawerView) {
                super.onDrawerClosed(drawerView);
            }

            @Override
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                hideKeyboard();
            }
        };
        mDrawer.addDrawerListener(mDrawerToggle);
        mDrawerToggle.syncState();
        setupDrawerNavMenu();
        setupBottomNavMenu();

        getViewDataBinding().mainViewPager.setAdapter(new MainViewPagerAdapter(
                getSupportFragmentManager(), MainActivity.this));
        mTabLayout.setupWithViewPager(getViewDataBinding().mainViewPager);
    }

    private void setupBottomNavMenu() {
        getViewDataBinding().bottomNavigation.setOnNavigationItemSelectedListener(this);
    }

    private void setupDrawerNavMenu() {
        NavHeaderMainBinding navHeaderMainBinding = DataBindingUtil.inflate(getLayoutInflater(),
                R.layout.nav_header_main, getViewDataBinding().navigationView, false);
        getViewDataBinding().navigationView.addHeaderView(navHeaderMainBinding.getRoot());
        navHeaderMainBinding.setViewModel(getViewModel());

        getViewDataBinding().navigationView.setNavigationItemSelectedListener(
                item -> {
                    mDrawer.closeDrawer(GravityCompat.START);
                    switch (item.getItemId()) {
                        case R.id.navItemAbout:
                            showFragment(AboutFragment.newInstance());
                            return true;
                        case R.id.navItemRelease:
                            showFragment(ArticleFragment.newInstance(APP_RELEASE_PAGE));
                            return true;
                        default:
                            return false;
                    }
                });
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(OpenArticleEvent event) {
        showFragment(ArticleFragment.newInstance(event.getUrl()));
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(OpenForumEvent event) {
        Properties prop = new Properties();
        prop.setProperty(PROP_FORUM_NAME, event.getTitle());
        EventHelperKt.sendKVEvent(EVENT_OPEN_FORUM, prop);
        showFragment(ForumArticlesFragment.newInstance(event.getTitle(), event.getUrl()));
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(OpenNotificationsPageEvent event) {

    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(NewMessageEvent event) {
        NotificationInfo info = event.getNotificationInfo();
        if (info.isNotEmpty()) {
            ToastUtilsKt.showMessage(getString(R.string.notification_arrival));
        }
    }

    private void showFragment(BaseFragment fragment) {
        ActivityUtils.addFragmentInActivity(getSupportFragmentManager(),
                fragment,
                R.id.clRootView);
    }

    @Override
    public boolean onFragmentDetached(String tag) {
        if (tag != null && tag.startsWith(TAG_PREFIX)) {
            return super.onFragmentDetached(tag);
        } else {
            return false;
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        menu.clear();
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.search, menu);
        // Get the SearchView and set the searchable configuration
//        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        SearchView searchView = (SearchView) menu.findItem(R.id.search).getActionView();
        searchView.setSubmitButtonEnabled(true);
//        // Assumes current activity is the searchable activity
//        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {

            @Override
            public boolean onQueryTextSubmit(String query) {
                hideKeyboard();
                Context context = mainActivityWeakReference.get();
                if (context != null) {
                    Intent intent = new Intent(context, SearchActivity.class);
                    intent.putExtra(URL_KEY, String.format(SEARCH_URL, query));
                    context.startActivity(intent);
                }
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.search) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void unlockDrawer() {
        if (mDrawer != null) {
            mDrawer.setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED);
        }
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
        if (getViewDataBinding().bottomNavigation.getSelectedItemId() == menuItem.getItemId()) {
            return false;
        }
        while (!mFragmentTags.isEmpty()
                && mFragmentTags.peek().startsWith(TAG_PREFIX)) {
            onFragmentDetached(mFragmentTags.pop());
        }
        switch (menuItem.getItemId()) {
            case R.id.action_favorite:
                showFragment(FavoriteArticlesFragment.newInstance());
                break;
            case R.id.action_forum_navigation:
                showFragment(ForumNavigationFragment.newInstance());
                break;
            case R.id.action_about_me:
                showFragment(MeFragment.newInstance());
                break;
            case R.id.action_home:
            default:
                break;
        }
        return true;
    }

    private static final int RW_EXTERNAL_STORAGE_CODE = 1111;
    private void requestPermissions() {
        String[] perms = {
                Manifest.permission.WRITE_EXTERNAL_STORAGE,
                Manifest.permission.READ_EXTERNAL_STORAGE
        };
        if (!EasyPermissions.hasPermissions(this, perms)) {
            EasyPermissions.requestPermissions(this, getString(R.string.permission_request), RW_EXTERNAL_STORAGE_CODE, perms);
        }
    }

    @Override
    public void onPermissionsGranted(int requestCode, List<String> perms) {
        ToastUtilsKt.showMessage(R.string.permission_grant);
    }

    @Override
    public void onPermissionsDenied(int requestCode, List<String> perms) {
        if (requestCode == RW_EXTERNAL_STORAGE_CODE) {
            ToastUtilsKt.showMessage(R.string.permission_denied);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NotNull String[] permissions, @NotNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this);
    }

    public static class MainViewPagerAdapter extends FragmentPagerAdapter {

        private List<TabModel> tabModels;

        MainViewPagerAdapter(FragmentManager fm, Context context) {
            super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
            tabModels = new ArrayList<>();
            tabModels.add(new TabModel(context.getString(R.string.tab_title_hot), "hot"));
            tabModels.add(new TabModel(context.getString(R.string.tab_title_tech), "tech"));
            tabModels.add(new TabModel(context.getString(R.string.tab_title_digest), "digest"));
            tabModels.add(new TabModel(context.getString(R.string.tab_title_new_thread), "newthread"));
        }

        @Override
        @NonNull
        public Fragment getItem(int position) {
            return ArticlesFragment.newInstance(tabModels.get(position).getUrl());
        }

        @Override
        public int getCount() {
            return tabModels.size();
        }

        @Nullable
        @Override
        public CharSequence getPageTitle(int position) {
            return tabModels.get(position).getTitle();
        }
    }

}

File: app/src/main/java/top/easelink/lcg/ui/main/view/MainActivity.kt
package top.easelink.lcg.ui.main.view

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.view.GravityCompat
import androidx.fragment.app.Fragment
import com.google.android.material.bottomnavigation.BottomNavigationItemView
import com.google.android.material.bottomnavigation.BottomNavigationMenuView
import com.google.android.material.bottomnavigation.BottomNavigationView
import kotlinx.android.synthetic.main.activity_main.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import top.easelink.framework.topbase.TopActivity
import top.easelink.framework.topbase.TopFragment
import top.easelink.framework.utils.addFragmentInActivity
import top.easelink.framework.utils.popBackFragmentUntil
import top.easelink.lcg.BuildConfig
import top.easelink.lcg.R
import top.easelink.lcg.config.AppConfig
import top.easelink.lcg.event.*
import top.easelink.lcg.ui.main.about.view.AboutFragment
import top.easelink.lcg.ui.main.article.view.ArticleFragment
import top.easelink.lcg.ui.main.articles.view.ForumArticlesFragment.Companion.newInstance
import top.easelink.lcg.ui.main.discover.view.DiscoverFragment
import top.easelink.lcg.ui.main.forumnav.view.ForumNavigationFragment
import top.easelink.lcg.ui.main.largeimg.view.LargeImageDialog
import top.easelink.lcg.ui.main.me.view.MeFragment
import top.easelink.lcg.ui.main.message.view.MessageFragment
import top.easelink.lcg.ui.main.model.*
import top.easelink.lcg.ui.main.recommand.view.RecommendFragment
import top.easelink.lcg.ui.setting.view.SettingActivity
import top.easelink.lcg.ui.webview.view.HalfScreenWebViewFragment
import top.easelink.lcg.ui.webview.view.WebViewActivity
import top.easelink.lcg.utils.WebsiteConstant.SERVER_BASE_URL
import top.easelink.lcg.utils.showMessage
import java.util.*
import kotlin.system.exitProcess

class MainActivity : TopActivity(), BottomNavigationView.OnNavigationItemSelectedListener {
    private var lastBackPressed = 0L
    private var bubbleView: View? = null

    override fun onBackPressed() {
        if (mFragmentTags.isNotEmpty() && mFragmentTags.size > 1) {
            while (onFragmentDetached(mFragmentTags.pop())) {
                syncBottomViewNavItemState()
                return
            }
        }
        if (System.currentTimeMillis() - lastBackPressed > 2000) {
            Toast.makeText(this, R.string.app_exit_tip, Toast.LENGTH_SHORT).show()
            lastBackPressed = System.currentTimeMillis()
        } else {
            finish()
            exitProcess(0)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        EventBus.getDefault().register(this)
        setupDrawer(toolbar)
        setupBottomNavMenu()
        showFragment(RecommendFragment::class.java)
        checkPermission()
    }

    override fun onDestroy() {
        super.onDestroy()
        EventBus.getDefault().unregister(this)
    }

    private fun setupDrawer(toolbar: Toolbar) {
        val header = layoutInflater.inflate(R.layout.nav_header, navigation_view, false)
        navigation_view.addHeaderView(header)
        app_version.text = BuildConfig.VERSION_NAME
        setSupportActionBar(toolbar)
        val toggle = ActionBarDrawerToggle(
            this,
            drawer_view,
            toolbar,
            R.string.open_drawer,
            R.string.close_drawer
        )
        drawer_view.addDrawerListener(toggle)
        toggle.syncState()
        navigation_view.setNavigationItemSelectedListener { item: MenuItem ->
            drawer_view.closeDrawer(GravityCompat.START)
            when (item.itemId) {
                R.id.nav_item_about -> {
                    showFragment(AboutFragment())
                    true
                }
                R.id.nav_item_release -> {
                    onMessageEvent(OpenArticleEvent(AppConfig.getAppReleaseUrl()))
                    true
                }
                R.id.nav_item_portal -> {
                    WebViewActivity.startWebViewWith(SERVER_BASE_URL, this)
                    true
                }
                R.id.nav_item_jrs -> {
                    WebViewActivity.startWebViewWith(AppConfig.getJrsUrl(), this)
                    true
                }
                R.id.nav_item_setting -> {
                    startActivity(Intent(this, SettingActivity::class.java))
                    true
                }
                else -> false
            }
        }
    }

    private fun setupBottomNavMenu() {
        bottom_navigation.setOnNavigationItemSelectedListener(this)
    }

    /**
     * manage the bottom navigation view item selected state
     * depends on current top fragment
     */
    private fun syncBottomViewNavItemState() {
        val view: BottomNavigationView = bottom_navigation
        try {
            view.setOnNavigationItemSelectedListener(null)
            when (mFragmentTags.peek()) {
                RecommendFragment::class.java.simpleName -> {
                    view.selectedItemId = R.id.action_home
                }
                MessageFragment::class.java.simpleName -> {
                    view.selectedItemId = R.id.action_message
                }
                ForumNavigationFragment::class.java.simpleName -> {
                    view.selectedItemId = R.id.action_forum_navigation
                }
                MeFragment::class.java.simpleName -> {
                    view.selectedItemId = R.id.action_about_me
                }
            }
        } catch (ese: EmptyStackException) {
            view.selectedItemId = R.id.action_home
        } finally {
            view.setOnNavigationItemSelectedListener(this)
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: OpenArticleEvent) {
        sendEvent(EVENT_OPEN_ARTICLE)
        if (AppConfig.articleShowInWebView) {
            WebViewActivity.startWebViewWith(SERVER_BASE_URL + event.url, this)
        } else {
            showFragment(ArticleFragment.newInstance(event.url))
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: OpenForumEvent) {
        val prop = Properties()
        prop.setProperty(PROP_FORUM_NAME, event.title)
        sendKVEvent(EVENT_OPEN_FORUM, prop)
        showFragment(newInstance(event.title, event.url, event.showTab))
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: NewMessageEvent) {
        val info = event.notificationInfo
        if (info.isNotEmpty()) {
            showMessage(getString(R.string.notification_arrival))
            showBubbleView(PRIVATE_MESSAGE_POS)
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onShowFragmentEvent(clazz: Class<TopFragment>) {
        showFragment(clazz)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: OpenLargeImageViewEvent) {
        if (event.url.isNotEmpty()) {
            LargeImageDialog.newInstance(event.url).show(
                supportFragmentManager,
                LargeImageDialog::class.java.simpleName
            )
        } else {
            showMessage(R.string.tap_for_large_image_failed)
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onOpenHalfScreenWebViewEvent(event: OpenHalfWebViewFragmentEvent) {
        HalfScreenWebViewFragment.newInstance(event.html).show(supportFragmentManager, HalfScreenWebViewFragment::class.simpleName)
    }

    @Suppress("SameParameterValue")
    private fun showBubbleView(pos: Int) {
        // Try show badge on bottom nav bar
        val menuView = bottom_navigation.getChildAt(0) as? BottomNavigationMenuView
        val itemView = menuView?.getChildAt(pos) as? BottomNavigationItemView
        bubbleView = LayoutInflater.from(this).inflate(R.layout.menu_badge, menuView, false)
        itemView?.addView(bubbleView)
    }

    @Suppress("SameParameterValue")
    private fun removeBubbleView(pos: Int) {
        bubbleView?.let {
            // remove bubble view
            val menuView = bottom_navigation.getChildAt(0) as? BottomNavigationMenuView
            val itemView = menuView?.getChildAt(pos) as? BottomNavigationItemView
            itemView?.removeView(it)
            bubbleView = null
        }
    }

    private fun showFragmentWithTag(tag: String): Boolean {
        supportFragmentManager.findFragmentByTag(tag)?.let {
            if (popBackFragmentUntil(supportFragmentManager, tag)) {
                var d = mFragmentTags.search(tag)
                while (d > 1) {
                    mFragmentTags.pop()
                    d--
                }
                return true
            }
        }
        return false
    }

    private fun showFragment(clazz: Class<*>) {
        if (!showFragmentWithTag(clazz.simpleName)) {
            addFragmentInActivity(
                supportFragmentManager,
                clazz.newInstance() as Fragment,
                R.id.clRootView
            )
        }
    }

    private fun showFragment(fragment: Fragment) {
        addFragmentInActivity(
            supportFragmentManager,
            fragment,
            R.id.clRootView
        )
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        return if (bottom_navigation.selectedItemId == item.itemId) {
            false
        } else {
            when (item.itemId) {
                R.id.action_message -> {
                    removeBubbleView(PRIVATE_MESSAGE_POS)
                    MessageFragment::class.java
                }
                R.id.action_forum_navigation -> DiscoverFragment::class.java
                R.id.action_about_me -> MeFragment::class.java
                R.id.action_home -> RecommendFragment::class.java
                else -> { null }
            }?.let {
                showFragment(it)
                true
            }?: false
        }
    }


    private fun checkPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
            != PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                WRITE_EXTERNAL_CODE
            )
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == WRITE_EXTERNAL_CODE) {
            if (grantResults.isEmpty() || (grantResults.isNotEmpty() && grantResults[0] != PERMISSION_GRANTED)) {
                showMessage(R.string.permission_denied)
            }
        }
    }

    companion object {
        const val WRITE_EXTERNAL_CODE = 1

        private const val PRIVATE_MESSAGE_POS = 2
    }
}